# Лабораторная работа №3 

Субботин Максим 9382, вариант 18, в млите 48, дм 38

## Цель работы
Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

##  Решение диофантового уравнения.
2771x+2261y=-136

In [19]:
from sympy.solvers.diophantine import diophantine
from sympy import symbols
x,y,z = symbols("x, y, z", integer=True)
diophantine(2771*x + 2261*y + 136)

{(133*t_0 + 248, -163*t_0 - 304)}

x=133t+248

y=-163t-304

## Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.
Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 3,m = 33).
Шифрованное сообщение имеет вид (21; 13; 3; 18). Подберите закрытую часть ключа и прочитайте исходное слово.
Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “e” и “ъ”).

In [22]:
from sympy.crypto.crypto import rsa_private_key, decipher_rsa, encipher_rsa

alphabet="абвгдежзийклмнопрстуфхцчшщыьэюя"
key = rsa_private_key(33, 3)
word = [21, 13, 3, 18]

string = [ alphabet[decipher_rsa(i, key)-2] for i in word]
print(string)

['у', 'е', 'з', 'д']


Получается слово 'уезд'

## Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

Функция имеет вид: f(x,y,z)=((yz) v (x v z)) xor (y xor (xy))

В решении переменные A,B,C соответствуют переменным x,y,z

In [23]:
from sympy.logic.boolalg import to_dnf
from sympy.abc import A, B, C
to_dnf((B & C) | (A | C) ^ (B ^ (A & B)), True)

A | B | C

In [24]:
from sympy.logic.boolalg import to_cnf
to_cnf((B & C) | (A | C) ^ (B ^ (A & B)), True)

A | B | C